File: 5/solution.py
import sys
import math


def parse_bsp(bspString):
    row_start = 0
    row_end = 127
    col_start = 0
    col_end = 7

    for character in bspString[:7]:
        pivot = math.ceil((row_start + row_end)/2)
        if character == 'B':
            row_start = pivot
        else:
            row_end = pivot - 1

    for character in bspString[7:]:
        pivot = math.ceil((col_start + col_end) / 2)
        if character == 'R':
            col_start = pivot
        else:
            col_end = pivot
    return (row_start,
            col_start,
            (row_start * 8) + col_start)


def part1(lines):
    max_id = -1
    for line in lines:
        result = parse_bsp(line)
        if max_id < result[2]:
            max_id = result[2]
    return max_id


def part2(lines):
    seats = []
    for line in lines:
        seats.append(parse_bsp(line))
    seats.sort(key=lambda tup: tup[2])
    for i in range(0, len(seats)-1):
        if seats[i][2]+2 == seats[i+1][2]:
            return seats[i][2]+1

    return -1


try:
    f = open(sys.argv[1], "r")
except IOError:
    print("Error opening the file, try again")
    sys.exit(1)
with f:
    rawLines = f.readlines()
    f.close()
    # digitArray = [int(s) for s in lines]
    print(
        f"Part 1 answer:{part1(rawLines)} Part 2 answer: {part2(rawLines)}")

File: 5/solution.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"sort"
)

type seat struct {
	row, col, ID int
}

func parseBSP(bspString string) seat {
	rowStart := 0
	rowEnd := 127
	colStart := 0
	colEnd := 7

	for i := 0; i < 7; i++ {
		pivot := ((rowStart + rowEnd + 1) / 2)
		if string(bspString[i]) == "B" {
			rowStart = pivot
		} else {
			rowEnd = pivot - 1
		}
	}

	for i := 7; i < 10; i++ {
		pivot := ((colStart + colEnd + 1) / 2)
		if string(bspString[i]) == "R" {
			colStart = pivot
		} else {
			colEnd = pivot - 1
		}
	}

	return seat{
		rowStart,
		colStart,
		(rowStart * 8) + colStart}
}

func part1(lines []string) int {
	maxID := -1
	for _, line := range lines {
		if maxID < parseBSP(line).ID {
			maxID = parseBSP(line).ID
		}
	}
	return maxID
}

func part2(lines []string) int {
	var seats []seat
	for _, line := range lines {
		seats = append(seats, parseBSP(line))
	}
	sort.Slice(seats, func(i, j int) bool {
		return seats[i].ID < seats[j].ID
	})
	for i := 0; i < len(seats)-1; i++ {
		if seats[i].ID+2 == seats[i+1].ID {
			return seats[i].ID + 1
		}
	}
	return -1
}

func main() {
	flag.Parse()
	if len(flag.Args()) == 0 {
		log.Fatal("Provide the path to input")
	}
	file, err := os.Open(flag.Args()[0])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}

	fmt.Printf("Part 1 answer:%v\nPart 2 answer:%v\n", part1(lines), part2(lines))
}
